lib/pbench/server/api/resources/query_apis/iteration_samples.py

```diff
@@ -0,0 +1,245 @@
+    """
+
+    WHITELIST_AGGS_FIELDS = ["run", "sample", "iteration", "benchmark"]
+
+    def __init__(self, config: PbenchServerConfig, logger: Logger):
+        super().__init__(
+            config,
+            logger,
+            Schema(Parameter("run_id", ParamType.STRING, required=True),),
+        )
+
+    def preprocess(self, client_json: JSON) -> CONTEXT:
+        """
+        Query the Dataset associated with this run id, and determine whether the
+        request is authorized for this dataset.
+
+        If the user has authorization to read the dataset, return the Dataset
+        object as CONTEXT so that the postprocess operation can proceed to get all
+        the iteration aggregations.
+        """
+        run_id = client_json.get("run_id")
+
+        # Query the dataset using the given run id
+        dataset = Dataset.query(md5=run_id)
+        if not dataset:
+            self.logger.error(f"Dataset with Run ID {run_id!r} not found")
+            abort(HTTPStatus.NOT_FOUND, message="Dataset not found")
+
+        owner = User.query(id=dataset.owner_id)
+        if not owner:
+            self.logger.error(
+                f"Dataset owner ID { dataset.owner_id!r} cannot be found in Users"
+            )
+            abort(HTTPStatus.INTERNAL_SERVER_ERROR, message="Dataset owner not found")
+
+        # For Iteration samples, we check authorization against the ownership of the
+        # dataset that was selected rather than having an explicit "user"
+        # JSON parameter. This will raise UnauthorizedAccess on failure.
+        self._check_authorization(owner.username, dataset.access)
+
+        # The dataset exists, and authenticated user has enough access so continue
+        # the operation with the appropriate CONTEXT.
+        return {"dataset": dataset, "run_id": run_id}
+
+
+class IterationSampleNamespaces(IterationSamples):
+    """
+    Iteration samples API that returns only unique namespaces for each
+    whitelisted result-data-sample subdocuments.
+
+    This class inherits the common IterationSamples class and builds an
+    aggregated query against a user supplied run id to compile the filter header
+    which shows every unique field in the 'benchmark', 'sample', 'iterations'
+    and 'run' subdocuments.
+    """
+
+    def __init__(self, config: PbenchServerConfig, logger: Logger):
+        super().__init__(config, logger)
+
+    def get_keyword_fields(self, mappings):
+        fields = []
+
+        def recurse(prefix, mappings, result):
+            if "properties" in mappings:
+                for p, m in mappings["properties"].items():
+                    recurse(f"{prefix}{p}.", m, result)
+            elif mappings.get("type") == "keyword":
+                result.append(prefix[:-1])  # Remove the trailing dot, if any
+            elif mappings.get("type") == "text":
+                for f, v in mappings.get("fields", {}).items():
+                    recurse(f"{prefix}{f}.", v, result)
```

Comment:
Should this just be an `else`? I wonder if it's possible to have other "primary" field types with `keyword` "fields" underneath? Like `date`, `ip`, etc.

Reply:
That is an interesting point and I think that make sense. We could make primary fields like IP a keyword by adding a `fields` underneath it.